content/post/graphql-with-go-and-postgresql.md

```diff
@@ -0,0 +1,377 @@
+   |                         <---------------------------+
+   |                         |                           |
+   |                         |                           |
+   |                         |                           |
+2. |                         |  SELECT * FROM users...   |
+   |                         +--------------------------->
+   |                         <---------------------------+
+   |                         |                           |
+   |                         |                           |
+   |                         |                           |
+3. |                         | SELECT * FROM comments... |
+   |                         +--------------------------->
+   |                         <---------------------------+
+   |                         |                           |
+   |                         |                           |
+   |                         |                           |
+4. |                         | SELECT * FROM followers.. |
+   |                         +--------------------------->
+   |                         <---------------------------+
+   <-------------------------+                           |
+   |                         |                           |
+   +                         +                           +
+```
+_Figure 2_
+
+_Listing 5_ contains the single GraphQL query which will fetch all the data needed to render the blog post.
+
+```graphql
+{
+  user(id: 1) {
+    email
+    followers
+    post(id: 1) {
+      title
+      body
+      comments {
+        id
+        title
+        user {
+          id
+          email
+        }
+      }
+    }
+  }
+}
+```
+_Listing 5_
+
+The number of queries to the database, for this scenario, is deliberately identical, but the number of http requests to the API server has been reduced to just one. We argue that the http requests over the Internet are the most costly in this type of application.
+
+The backend doesn't have to be designed differently to start reaping the benefits of GraphQL, transitioning from REST to GraphQL can be done incrementally. This allows to measure performance improvements and optimize. From this point, the API developer can start to optimize (potentially merge) SQL queries to improve performance. The opportunity for caching is greatly increased, both on the database and API levels.
+
+Abstractions on top of SQL (for instance ORM layers) usually have to contend with the `n+1` problem. In step `4.` of the REST example, a client could have had to request the follower status for the author of each comment in separate requests. This is because in REST there is no standard way of expressing relationships between more that two resources, whereas GraphQL was designed to prevent this problem by using nested queries. Here, we cheat, by fetching all the followers of the user and deferring to the client the logic of determining users who commented and also follow the author.
```

Comment:
- more **than** two resources
- no comma: we cheat by
- determining **the** users
- **following** the author